xds/internal/xdsclient/client_refcounted.go

```diff
@@ -35,41 +34,33 @@ const (
 )
 
 var (
-	// This is the client returned by New(). It contains one client implementation,
-	// and maintains the refcount.
-	singletonMu     sync.Mutex
-	singletonClient *clientRefCounted
-
 	// The following functions are no-ops in the actual code, but can be
 	// overridden in tests to give them visibility into certain events.
-	singletonClientImplCreateHook = func() {}
-	singletonClientImplCloseHook  = func() {}
+	xdsClientImplCreateHook = func(name string) {}
+	xdsClientImplCloseHook  = func(name string) {}
 )
 
-// To override in tests.
-var bootstrapNewConfig = bootstrap.NewConfig
-
-func clientRefCountedClose() {
-	singletonMu.Lock()
-	defer singletonMu.Unlock()
+func clientRefCountedClose(name string) {
```

Comment:
Optional: this function doesn't necessarily close the clientImpl for the name. It decrements the ref and if it's the last ref, it closes/removes it from global map. Should this be called clientRefCountedDecrementRef instead? I see it was already called this.

Reply:
I don't feel strongly. Is it because semantically a ref counted thing is known to do this behind the scenes?

Reply:
Yes, this would have simply been `close` or `clientClose`. But because we have different client-ish objects here, the `XDSClient` interface, the `clientImpl` which is the implementation, and this one, which is the ref counted wrapper, the name ended up containing the `RefCounted` part.

On second thoughts though, we might be able to get rid of this type completely and move the reference counting functionality to either the `clientImpl` type or we could come up with a generic reference counted type that can wrap anything. I will file an issue on github to explore this further (don't want to do it as part of this PR, which is already huge).

Reply:
The latter sounds good. I remember from C++ they have a ptr type that essentially does ref counting/freeing for you so it would be a type in Go that serves similar functionality.